source/shared/cpp/ObjectModel/MarkDownParser.cpp

```diff
@@ -0,0 +1,477 @@
+int MarkDownParser::AdjustEmphasisCounts(int leftOver, std::list<Emphasis>::iterator left_itr, 
+        std::list<Emphasis>::iterator right_itr)
+{
+    int delimCnts = 0;
+    if (leftOver >= 0)
+    {
+        delimCnts = left_itr->m_emphCnts - leftOver;
+        left_itr->m_emphCnts = leftOver;
+        right_itr->m_emphCnts = 0;
+    }
+    else
+    {
+        delimCnts = left_itr->m_emphCnts;
+        right_itr->m_emphCnts = leftOver * (-1);
+        left_itr->m_emphCnts = 0;
+    }
+    return delimCnts;
+}
+
+// Following the rules speicified in CommonMark (http://spec.commonmark.org/0.27/)
+// It generally supports more stricker version of the rules
+// push left delims to stack, until matching right delim is found,
+// update emphasis counts and type to build string after search is complete
+std::string MarkDownParser::TransformToHtml()
+{
+    GenSymbolTable();
+    if (!m_leftLookUpTable.size() || !m_rightLookUpTable.size())
+    {
+        return "<p>" + m_text + "</p>";
+    }
+
+    std::vector<std::list<Emphasis>::iterator> DFSSearchStack;
+    auto top_left = m_leftLookUpTable.begin(), top_right = m_rightLookUpTable.begin();
+    auto curr_left_delim = top_left, right_token_idx = top_right;
+    int lookBehind = (m_tokenizedString[top_left->m_idx][0] == '*')? Asterisk : Underscore; 
+    DFSSearchStack.push_back(top_left++);
+
+    while (!DFSSearchStack.empty() && top_right != m_rightLookUpTable.end())
+    {
+        // check if left most emphasis passed a new right emphasis delimiter
+        if (top_left != m_leftLookUpTable.end() && top_left->m_idx < top_right->m_idx)
+        {
+            DFSSearchStack.push_back(top_left);
+            lookBehind |= (m_tokenizedString[top_left->m_idx][0] == '*')? Asterisk : Underscore; 
+            ++top_left;
+        }
+        else
+        {
+            curr_left_delim = DFSSearchStack.back();
+            // because of rule #9 & #10 and multiple of 3 rule, left delim can jump ahead of right delim,
+            // so need to check this condition.
+            if (curr_left_delim->m_idx > top_right->m_idx)
+            {
+                top_right++;
+                continue;
+            }
+            int delimCnts = 0, leftOver = 0;
+            DelimiterType curr_left_delim_type = (m_tokenizedString[curr_left_delim->m_idx][0] == '*')? Asterisk : Underscore; 
+            DelimiterType curr_right_delim_type = (m_tokenizedString[top_right->m_idx][0] == '*')? Asterisk : Underscore; 
+
+            // found the left and right delimiter, gen string
+            if (curr_left_delim_type == curr_right_delim_type)
+            {
+                // rule #9 & #10, sum of delim cnts can't be multipe of 3 
+                if (!((curr_left_delim->m_emphCnts + top_right->m_emphCnts) % 3))
+                {
+                      if (m_tokenizedString.size() != top_right->m_idx + 1)
+                      {
+                          // check adjcent char is space
```

Comment:
>adjcent [](start = 35, length = 7)

Adjacent

Reply:
thanks!

---
In reply to: [154774189](https://github.com/Microsoft/AdaptiveCards/pull/1003#discussion_r154774189) [](ancestors = 154774189)